### PR TITLE
output .links.json

### PR DESCRIPTION
**Pull request by live1206**
https://dev.azure.com/ceapex/Engineering/_workitems/edit/123107/

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5124)

### CODE REVIEW

File: docs/specs/moniker.yml [hunk 1]
```diff
@@ -1407,3 +1407,72 @@ outputs:
...
+    }
+  23d205a5/docs/c.json: |
+    {
+      "conceptual": "<p>some text</p>",
+      "monikers": ["netcore-2.0","netcore-2.1"]
+    }
+  .links.json: |
+    {
+      "links":[
+        {"source_url":"v1/b.md","target_url":"/docs/b"},
```
Comment (OsmondJiang) [inline docs/specs/moniker.yml]: >"source_url":"v1/b.md" [](start = 9, length = 22)

source url should be the absolute link of source file

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 2]
```diff
@@ -94,6 +98,15 @@ internal class DependencyResolver
                     referencingFile, isSelfBookmark ? relativeToFile : file, fragment, isSelfBookmark, path);
             }
 
+            // ignore bookmark reference for FileLinkMap
+            if (!isSelfBookmark
+                && UrlUtility.FragmentToDependencyType(fragment) != DependencyType.Bookmark
+                && !string.IsNullOrEmpty(link)
+                && file != null)
+            {
+                _fileLinkMapBuilder.AddFileLink(path, referencingFile, link);
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: >path, referencingFile [](start = 48, length = 21)

referencingFile is enough here

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 3]
```diff
@@ -94,6 +98,15 @@ internal class DependencyResolver
                     referencingFile, isSelfBookmark ? relativeToFile : file, fragment, isSelfBookmark, path);
             }
 
+            // ignore bookmark reference for FileLinkMap
+            if (!isSelfBookmark
+                && UrlUtility.FragmentToDependencyType(fragment) != DependencyType.Bookmark
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: >&& UrlUtility.FragmentToDependencyType(fragment) != DependencyType.Bookmark [](start = 16, length = 75)

why ignore link with bookmarks?

File: src/docfx/build/link/FileLinkItem.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Docs.Build
+{
+    [JsonObject(NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
+    internal class FileLinkItem
+    {
+        public string SourceUrl { get; set; }
+
+        public string SourceMonikerGroup { get; set; }
```
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/build/link/FileLinkItem.cs]: Thought we agreed on the design spec?
https://github.com/dotnet/docfx/blob/v3/docs/designs/link-output.md

File: docs/specs/moniker.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (live1206) [reply docs/specs/moniker.yml]: outdated

File: docs/specs/moniker.yml [hunk 5]
```diff
@@ -1407,3 +1407,72 @@ outputs:
         "9d4e15fd":["netcore-1.0","netcore-1.1"]
       }
     }
+---
+# output .links.json
+inputs:
```
[the remark this answers is not in the document]
Comment (live1206) [reply docs/specs/moniker.yml]: The TOC links do not contain monikers, would that be enough?

File: docs/specs/moniker.yml [hunk 6]
```diff
@@ -1407,3 +1407,72 @@ outputs:
         "9d4e15fd":["netcore-1.0","netcore-1.1"]
       }
     }
+---
```
[the remark this answers is not in the document]
Comment (live1206) [reply docs/specs/moniker.yml]: sounds good

File: src/docfx/build/link/FileLinkItem.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/build/link/FileLinkItem.cs]: They gave up the new requirement at last

File: docs/specs/output.yml [hunk 7]
```diff
@@ -412,3 +412,65 @@ outputs:
...
+    uid: a
+    title: netcore
+    monikerRange: '> netcore-1.0'
+    ---
+    Moniker: netcore-1.1
+  docs/v1/b.md: |
+    Link to @a
+    Link to [some text](../v2/c.md)
+    Link to [not existed](c.md)
+    Link to [absolute link](/docs/c)
```
Comment (OsmondJiang) [inline docs/specs/output.yml]: >to [](start = 9, length = 2)

what about link to empty like [empty]()

Comment (live1206) [reply docs/specs/output.yml]: We would not output this since a non-null/empty link check before adding them

**Comment by live1206** [comment]
> hold off auto merge

integration test is failing, it will not auto merge

File: src/docfx/build/link/FileLinkMapBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            Debug.Assert(monikerProvider != null);
+            _monikerProvider = monikerProvider;
+        }
+
+        public void AddFileLink(Document file, string targetUrl)
+        {
+            Debug.Assert(file != null);
+            Debug.Assert(targetUrl != null);
+
+            var (_, monikers) = _monikerProvider.GetFileLevelMonikers(file);
```
Comment (yufeih) [inline src/docfx/build/link/FileLinkMapBuilder.cs]: This line should have throw for TOC files due to https://github.com/dotnet/docfx/pull/5101 @928PJY

**Comment by yufeih** [comment]
I talked with qianhao and xiaofu, we agree that `monikerName` is better then `monikerGroupName` as a contract. Still there are some moniker related details that needs close (The devil is in the details)

File: src/docfx/build/link/FileLinkMapBuilder.cs (hunk 8 above)
Comment (928PJY) [reply src/docfx/build/link/FileLinkMapBuilder.cs]: No, that PR has been abandoned, and we decide to use the monkers from config and file metadata as the monikers of toc

---
In reply to: [327410727](https://github.com/dotnet/docfx/pull/5124#discussion_r327410727) [](ancestors = 327410727)

Comment (live1206) [reply src/docfx/build/link/FileLinkMapBuilder.cs]: Adding tests to cover this scenario as well

Comment (yufeih) [inline src/docfx/build/link/FileLinkMapBuilder.cs]: do not ignore errors

**Pull request merged**